DNN Platform/DotNetNuke.Web/Api/Auth/ApiTokens/ApiTokenController.cs

```diff
@@ -0,0 +1,282 @@
+            return string.IsNullOrEmpty(authorization) ? (null, null) : this.ValidateAuthorizationValue(authorization);
+        }
+
+        /// <inheritdoc />
+        public void SetApiTokenForRequest(ApiToken token)
+        {
+            HttpContext.Current.Items["ApiToken"] = token;
+        }
+
+        /// <inheritdoc />
+        public ApiToken GetCurrentThreadApiToken()
+        {
+            if (HttpContext.Current != null && HttpContext.Current.Items["ApiToken"] is ApiToken token)
+            {
+                return token;
+            }
+
+            return null;
+        }
+
+        /// <inheritdoc />
+        public SortedDictionary<string, ApiTokenAttribute> ApiTokenKeyList(ApiTokenScope scope, string locale)
+        {
+            var res = new SortedDictionary<string, ApiTokenAttribute>();
+            var typeLocator = new TypeLocator();
+            var attributes = typeLocator.GetAllMatchingTypes(
+                t => t != null &&
+                     t.IsClass &&
+                     !t.IsAbstract &&
+                     t.IsVisible)
+                .SelectMany(x => x.GetMethods())
+                .SelectMany(m => m.GetCustomAttributes(typeof(ApiTokenAuthorizeAttribute), false))
+                .Cast<ApiTokenAuthorizeAttribute>()
+                .Where(a => a.Scope <= scope);
+
+            foreach (var attr in attributes)
+            {
+                var key = attr.Key.ToLowerInvariant();
+                var k = attr.Scope.ToString() + key;
+                if (!res.ContainsKey(k))
+                {
+                    var name = DotNetNuke.Services.Localization.Localization.GetString(attr.Key + ".Text", attr.ResourceFile, locale);
+                    var description = DotNetNuke.Services.Localization.Localization.GetString(attr.Key + ".Help", attr.ResourceFile, locale);
+                    res.Add(k, new ApiTokenAttribute((int)attr.Scope, key, name, description));
+                }
+            }
+
+            return res;
+        }
+
+        /// <inheritdoc />
+        public IPagedList<ApiToken> GetApiTokens(ApiTokenScope scope, bool includeNarrowerScopes, int portalId, int userId, ApiTokenFilter filter, string apiKey, int pageIndex, int pageSize)
+        {
+            return this.apiTokenRepository.GetApiTokens(scope, includeNarrowerScopes, portalId, userId, filter, apiKey, pageIndex, pageSize);
+        }
+
+        /// <inheritdoc />
+        public string CreateApiToken(int portalId, string tokenName, ApiTokenScope scope, DateTime expiresOn, string apiKeys, int userId)
+        {
+            if (scope == ApiTokenScope.Host)
+            {
+                portalId = Null.NullInteger;
+            }
+
+            string newToken;
+            using (var generator = RandomNumberGenerator.Create())
+            {
+                var tokenBytes = new byte[32];
+                generator.GetBytes(tokenBytes);
+                newToken = Convert.ToBase64String(tokenBytes);

```

Comment:
I think we need to use the same conversion here (from `byte[]` to `string`) that we're using during hashing.

```suggestion
                newToken = EncodeBase64(tokenBytes);
```